src/main/java/uk/gov/hmcts/reform/bailcaseapi/domain/handlers/presubmit/UploadDocumentsHandler.java

```diff
@@ -0,0 +1,141 @@
+                        : isHomeOffice(bailCase)
+                ? "Home Office"
+                : "");
+
+        if (maybeDocument.isPresent()) {
+            List<DocumentWithMetadata> document =
+                maybeDocument
+                    .orElseThrow(() -> new IllegalStateException("document is not present"))
+                    .stream()
+                    .map(IdValue::getValue)
+                    .map(doc -> documentReceiver.tryReceive(doc, DocumentTag.UPLOAD_DOCUMENT, suppliedBy))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toList());
+
+            BailCaseFieldDefinition appropriateCollectionForUserRole =
+                selectAppropriateDocumentsCollection(bailCase, suppliedBy);
+
+            Optional<List<IdValue<DocumentWithMetadata>>> maybeExistingDocuments =
+                bailCase.read(appropriateCollectionForUserRole);
+
+            final List<IdValue<DocumentWithMetadata>> existingHomeOfficeDocuments =
+                maybeExistingDocuments.orElse(Collections.emptyList());
+
+            List<IdValue<DocumentWithMetadata>> allDocuments =
+                documentsAppender.append(existingHomeOfficeDocuments, document);
+
+            bailCase.write(appropriateCollectionForUserRole, allDocuments);
+
+            bailCase.clear(UPLOAD_DOCUMENTS);
+            bailCase.clear(UPLOAD_DOCUMENTS_SUPPLIED_BY);
+        }
+        return new PreSubmitCallbackResponse<>(bailCase);
+    }
+
+    private boolean isLegalRep(BailCase bailCase) {
+        return bailCase.read(BailCaseFieldDefinition.IS_LEGAL_REP, YesOrNo.class).map(flag -> flag.equals(
+            YesOrNo.YES)).orElse(false);
+    }
+
+    private boolean isHomeOffice(BailCase bailCase) {
+        return bailCase.read(BailCaseFieldDefinition.IS_HOME_OFFICE, YesOrNo.class).map(flag -> flag.equals(
+            YesOrNo.YES)).orElse(false);
+    }
+
+    private BailCaseFieldDefinition selectAppropriateDocumentsCollection(BailCase bailCase, String suppliedBy) {
+        BailCaseFieldDefinition appropriateDocumentsCollection;
+
+        if (isHomeOffice(bailCase)) {
+            appropriateDocumentsCollection = HOME_OFFICE_DOCUMENTS_WITH_METADATA;
+        } else if (isLegalRep(bailCase)) {
+            appropriateDocumentsCollection = APPLICANT_DOCUMENTS_WITH_METADATA;
+        } else {
+            appropriateDocumentsCollection = suppliedBy.equals(SUPPLIED_BY_APPLICANT)
+                                             || suppliedBy.equals(SUPPLIED_BY_LEGAL_REPRESENTATIVE)
+                ? APPLICANT_DOCUMENTS_WITH_METADATA
+                : suppliedBy.equals(SUPPLIED_BY_HOME_OFFICE)
+                ? HOME_OFFICE_DOCUMENTS_WITH_METADATA
+                : null;
+        }
+
+        if (appropriateDocumentsCollection == null) {
+            throw new IllegalStateException("Impossible to determine which collection to append the document to");
```

Comment:
Can you please change the error to: Unable to determine the supplier of the document,

Reply:
Renamed